### PR TITLE
[enh] add checker

### LINKED ISSUE #1559: Embedded searx-checker
I've updated searx-checker to produce an html output, see https://searx.space
Source code based on work of @kvch : https://github.com/searx/searx-checker

It shows which engines currently doesn't work (warning : some engine may work / not work from time to time).

The thing that would be nice is to be able test the different searx instances.
Example : the first instance in https://stats.searx.xyz/ is https://searx.at
This instance answers quickly with the last version searx and good security configuration. 
But this instance doesn't work : whatever the engine is, whatever the query is, it always shows the same error message. It would be nice to show this information on stats.searx.xyz

Using searx-checker doesn't work : most the searx engines disable and severly limit json output.

A solution is to embed searx-checker inside searx. 
Every day searx makes a self test of all the configured engines. It can be done easily if each engine contains a list of queries that are supposed to provide answers (with a default one is not configured).
The result would be available on a /status URL : see https://searx.space/output/localhost.json for an example.

Then stats.searx.xyz can display a sum up of this information : 
* on the current view : number of working engine per instance (perhaps with a detail per category).
* another view (to create) : number of working instance per engine.

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?

This PR:
* implements https://github.com/searx/searx/issues/1559
* take [searx-checker](https://github.com/searx/searx-checker) over.
* adds a new dependency: [pycld3](https://github.com/bsolomon1124/pycld3) (implementation of `has_lang`)

----

An engine can define a new global variable `tests`.
This variable can be defined in the code ( `searx/engines/*.py` or ) or in `settings.yml`.

Content example written in YAML :
```yaml
- matrix:
    # 3 requests search for "test" in english, pages 1, 2 and 3.
    query: test
    pageno: [1, 2, 3]
    lang: en
- result_container:
    # check each page is not empty
    - not_empty
    # check the results are in english
    - [ "has_lang", "en" ]
- test:
    # check each result is unique (page 1 is not page 2)
    - unique_results
```

* each entry in `result_container` call is a method name of `searx.search.checker.ResultContainerTests`.
* each entry in `test` call is a method name of `searx.search.checker.CheckerTests`.

There are default tests:
* see `searx.search.processors.abstract.EngineProcessor`
   * the method `get_tests` return the value define in the engine (or in settings.yml)
   * if there is no value, then some default tests are returned by the `get_default_tests` method.
* `searx.search.processors.online.OnlineProcessor.get_default_tests` contains some generic tests.
	* The tests are variable according to what the engine support. For example there is a test about paging when the engine support this feature)
	* the 'rosebud' query idea comes from https://searx.neocities.org/changelog.html
* `searx.search.processors.online_currency.OnlineCurrencyProcessor.get_default_tests` contains on specific test.
* `searx.search.processors.online_dictionary.OnlineDictionaryProcessor.get_default_tests` contains on specific test.

----

For the now, there is a dedicated main program: `python -u -m searx.search.checker`

Output: https://gist.github.com/dalf/3457519d264384bfc7f234b7d8ec91b3 (slightly modified using `sort`)

It is possible to specify some engine name:
`python -u -m searx.search.checker "bing images" "google images"'

----

What is missing:
* [x] fix CI
* [x] check install with arch
* [x] check install with fedora
* [x] check install with centos
* [x] scheduling & API (scheduling might be clunky as long searx relies on uwsgi)
* [x] proper tool to help development
* [x] infobox checks
* [x] make sure `_is_url_image` works as intended.
* [x] adjust the tests for different engines.

## Why is this change important?

For reference #1559

Currently if is difficult to know which engine is broken.

https://github.com/searx/searx/pull/2332 helps to know runtime error, but for example it doesn't detect if one these feature is broken: paging, time_range, safesearch.

This PR brings a easy way (?) to add tests, include for engines only define in settings.yml

## How to test this PR locally?

* For now `python -u -m searx.search.checker`

## Related issues

Close #1559

### CODE REVIEW

**Review by return42 — COMMENTED** [review]
My first test on fedora  ..

```
sudo -H ./utils/lxc.sh build searx-fedora31
sudo -H ./utils/lxc.sh cmd searx-fedora31 ./utils/searx.sh install packages
sudo -H ./utils/lxc.sh cmd searx-fedora31 make clean pyenvinstall
sudo -H ./utils/lxc.sh cmd searx-fedora31 make searx.checker
```

looks good.  I also tested `searx-archlinux` and `searx-centos7`

The `searx.checker` itself seems to work on all platforms without any issue in the installation procedure [(package additions in utils/searx.sh are OK)](https://github.com/searx/searx/pull/2419/commits/f0a40b4bbe77fa82f8d26d3c0d634dfb831dfbe6#diff-ffd563c55bf191e4f0f3bdc162add389733738c2e88c5a3e9c22d54585f7be69).

On some platforms searx.checker reports more errors and raises more exceptions compared to my very first test on the desktop of my ubu2004.  

There are issues like `SearxEngineTooManyRequestsException` and `CAPTCHA` which might related to the mass testing. I will have a deeper look for what the reasons are .. _coming soon_.

Here is the branch with additional commits, I tested on: https://github.com/return42/searx/commits/checker

**Comment by HLFH** [comment]
I have tested on ArchLinux as I maintain [searx-git on the AUR](https://aur.archlinux.org/packages/searx-git/).

In my *searx-git* PKGBUILD, I changed:
> source=(git+https://github.com/asciimoo/searx

to:
> source=(git+https://github.com/dalf/searx#branch=checker

And then, I tested the PR.

> python -u -m searx.search.checker
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/runpy.py", line 188, in _run_module_as_main
>     mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
>   File "/usr/lib/python3.9/runpy.py", line 147, in _get_module_details
>     return _get_module_details(pkg_main_name, error)
>   File "/usr/lib/python3.9/runpy.py", line 111, in _get_module_details
>     __import__(pkg_name)
>   File "/usr/lib/python3.9/site-packages/searx/search/checker/__init__.py", line 1, in <module>
>     from .impl import Checker
>   File "/usr/lib/python3.9/site-packages/searx/search/checker/impl.py", line 9, in <module>
>     import cld3
> ModuleNotFoundError: No module named 'cld3'

I installed the dependency *python-cld3-git*.

Output: https://gist.github.com/HLFH/7779f70d2b091b0066384c0d1e2e9f0e

File: searx/search/checker/__main__.py
Comment (dalf) [reply searx/search/checker/__main__.py]: What the verbose mode would do ?

Comment (return42) [reply searx/search/checker/__main__.py]: The condition `if sys.stdout.isatty():` prints "verbose" messages 

     print(BOLD_SEQ, 'Engine ', '%-30s' % name, RESET_SEQ, WHITE, ' Checking', RESET_SEQ)

only when the command is started with a tty output, not when piping the output to a file.

My suggestion was to add "-v" flag instead this tty-condition.  But this is nothing mandatory.

File: searx/search/checker/__main__.py
Comment (dalf) [reply searx/search/checker/__main__.py:94]: If there are different options / parameters, Makefile is not very convenient ?

`python -m searx.search.checker duckduckgo "google images"`

What about an entrypoint ?
https://github.com/searx/searx/blob/14a395a6718949015e53a0b593eaa2f28e43fd27/setup.py#L52-L56

**Comment by dalf** [comment]
To schedule the checker, there is a new module `searx.shared` : 
* `searx.shared.storage` allows to get / set `int` and `str`.
* `searx.shared.schedule(delay, func, *args)` schedules a `func(*args)` every `delay` seconds.

If searx detects uwsgi, searx uses uwsgi cache2:
* https://uwsgi-docs.readthedocs.io/en/latest/Caching.html
* https://uwsgi-docs.readthedocs.io/en/latest/PythonModule.html#cache-functions

But it requires to add this line to uwsgi.ini : 
`cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1`.

I think the uwsgi.ini is ossified (there are too many way to install searx), so if searx detect uwsgi but uwsgi.ini is not updated, the checker is disabled.

Without uwsgi, searx uses [`threading.Timer(delay, wrapper)`](https://docs.python.org/3.10/library/threading.html#timer-objects)

**Comment by dalf** [comment]
@HLFH  thank you for the test ! 

I hope the report is clear.

**Comment by dalf** [comment]
Usage cases of searx-checker command line:
```sh
$ make pyenvinstall
$ . ./local/py3/bin/activate

$ searx-checker google bing
Engine google                        Checking
Engine google                        OK
Engine bing                          Checking
Engine bing                          ErrorError ['paging: No result']

$ searx-checker -v google bing
Engine google                        Checking
Engine google                        OK
    found languages: en es et fr lb
Engine bing                          Checking
Engine bing                          Error
    found languages: en fr pt
    simple         : No result (query='computer' lang='all' pageno=1 safesearch=0 time_range=None)
    paging         : No result (query='news' lang='all' pageno=2 safesearch=0 time_range=None)
    paging         : No result (query='news' lang='all' pageno=3 safesearch=0 time_range=None)
    paging         : results are identitical for pageno=2 and pageno=3 (query='news' lang='all' safesearch=0 time_range=None)


$ searx-checker -v bing > result.txt
Engine bing                          Checking

$ cat result.txt
Engine bing                          Checking
Engine bing                          Error
    found languages: en
    simple         : No result (query='life' lang='all' pageno=1 safesearch=0 time_range=None)
    paging         : No result (query='news' lang='all' pageno=1 safesearch=0 time_range=None)
```

It is possible to list different engine either using their names or their shortcuts.
An empty list means all the engines.

@return42  I don't know how to make convenient using make.

File: searx/search/checker/__main__.py
Comment (dalf) [reply searx/search/checker/__main__.py]: The new version:
* always write to stdout "Engine ... checking"
* write to stderr "Engine ... checking" too when ` sys.stdout.isatty()`

See below https://github.com/searx/searx/pull/2419#issuecomment-756918650

**Comment by dalf** [comment]
Perhaps some engine require some adjustments, but I think the foundation is here to:
* run the checker automatically (result at `/stats/checker`.
   * Later searx-stats2 can fetch this URL (and `/stats/errors`).
   * Later it is possible to display the health status of the engines in the preferences.
* check / add the tests during the development phase of an engine.
* quickly see which engines don't work.

**Comment by HLFH** [comment]
I now use a searx private instance.

I tested in command line.
> ➜  sites-enabled searx-checker google bing
> Engine google                        Checking
> Engine google                        OK
> Engine bing                          Checking
> Engine bing                          ErrorError ['simple: No result', 'paging: No result', "paging: results are identitical for pageno=2 and pageno=3 (query='news' lang='all' safesearch=0 time_range=None)"]

And this new UI feature:
> Display the health status of the engines in the preferences

... this would be awesome! With this clearer engines report thanks to a neat UI, it will make things easier to submit and check on engines issues.

Shall I add the line `cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1` to my `/etc/uwsgi/vassals/searx.ini`? And do `systemctl restart uwsgi@searx.service`?
Or after like 10 mins without editing the file, my results on "/searx/stats/checker" will be generated instead of having "null"?
Do the results on "/searx/stats/checker" show the entry "checked_at" as well?

> I think the uwsgi.ini is ossified (there are too many way to install searx), so if searx detect uwsgi but uwsgi.ini is not updated, the checker is disabled.

Does the command line `searx-checker status` exist? So I can check if it is enabled or disabled?

**Comment by dalf** [comment]
> Shall I add the line cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1 to my /etc/uwsgi/vassals/searx.ini? And do systemctl restart uwsgi@searx.service?

Yes ! Without this line, the checker is disabled if you are using uwsgi.

Also adjust the `checker` section in settings.yml to your needs:
https://github.com/searx/searx/blob/d347d244c7f02104b35b406bcbf25c7362c5ad2d/searx/settings.yml#L100-L102

The checker starts after a random delay after searx has started. This delay is by default between 5 and 30 minutes (`1800` seconds).

After that, by default the checker runs every `86400` seconds (one day).
More precisely, it run every `864000` seconds, and for each run, it waits an additional random delay between `0` and `90000 - 86400` seconds (for the default configuration).

Notes:
* if there is no `checker.scheduling` section in settings.yml, then the checker is disabled.
* you can start the checker manually by sending a SIGUSR1 signal to any of the uwsgi process.
* if you are running searx in debug mode, the checker is disabled by default :
https://github.com/searx/searx/blob/d347d244c7f02104b35b406bcbf25c7362c5ad2d/searx/settings.yml#L97

File: searx/shared/shared_abstract.py
Comment (kvch) [inline searx/shared/shared_abstract.py]: Why don't you use Abstract Base Class here?

File: searx/search/checker/background.py
Comment (kvch) [inline searx/search/checker/background.py]: Nit: maybe instead of `status` we could name it `success`, so the boolean value makes more sense.

File: searx/shared/__init__.py
Comment (kvch) [inline searx/shared/__init__.py:31]: Why is it a dict? What else are you planning to store in `storage` that you need to have in the same key-value storage?

File: searx/search/checker/background.py
Comment (kvch) [inline searx/search/checker/background.py]: Nit: use the constant `CHECKER_RESULT` you have defined above in line 17.

File: searx/shared/__init__.py
Comment (dalf) [reply searx/shared/__init__.py:31]: * `ShareDict` encoding / decoding is simple : uwsgi deals with `bytes` not a Python object.
* the caller declares what kind of data is expected (not Python spirit I know, but here it is about cooperation between multiple process, I prefer something more declarative)

The alternative is to serialize the Python object using `pickle` or something equivalent, so `SharedDict`, so we can have `storage[key] = value` and `storage[key]`.

> What else are you planning to store in storage that you need to have in the same key-value storage?

Response times: https://github.com/searx/searx/pull/477 : here pickling each number is not very efficient.

Important note: the uwsgi implementation won't work with https://github.com/searx/searx/pull/2313

An alternative implementation can use [anonymous memory map](https://docs.python.org/3/library/mmap.html) which also provides `bytes` (without a hash table).

File: searx/shared/shared_abstract.py
Comment (dalf) [reply searx/shared/shared_abstract.py]: Fixed in this commit https://github.com/searx/searx/pull/2419/commits/eca200404d195fcfaf29a8b30f3cc7d2c7e4c021

File: searx/search/checker/background.py
Comment (dalf) [reply searx/search/checker/background.py]: Fixed in this commit https://github.com/searx/searx/pull/2419/commits/a7e980251b88e12167dd2012aa1e36c76657e6bf

Also answer @HLFH  question : 

> Does the command line searx-checker status exist? So I can check if it is enabled or disabled?

If the checker is disabled, `/stats/checker` returns:
```json
{
  "status": "disabled",
  "timestamp": 1610388000
}
```

* `timestamp` is round to hour. Basically here it shows when the instance was started. Is it an issue ? It can be rounded to day if necessary.
* `kill -SIGUSR1 ...` can still trigger the checker.

The status is:
* `ok` after an update.
* `error` if an exception stop the checker

File: searx/search/checker/background.py
Comment (dalf) [reply searx/search/checker/background.py]: Done in the commit https://github.com/searx/searx/pull/2419/commits/a7e980251b88e12167dd2012aa1e36c76657e6bf

File: searx/search/checker/background.py
Comment (HLFH) [inline searx/search/checker/background.py]: On my searx private instance _http://MYLOCALIP/searx/stats/checker_, I have:

{"status":"unknow","timestamp":1610398800}

Typo: should be "unknown".

Comment (HLFH) [reply searx/search/checker/background.py]: And when I try to refresh this _checker_ web page multiple times during the same minute, I can get various results: _error_, _unknow_, and sometimes the results.
I tried with in Chrome Incognito Window, and normal window.

Comment (dalf) [reply searx/search/checker/background.py]: It should be : 
* `unknow(n)` when searx starts.
* then `ok` after an success run of the checker (with or without engine errors)
* or `error` after an exception ( [background.py](https://github.com/searx/searx/blob/eca200404d195fcfaf29a8b30f3cc7d2c7e4c021/searx/search/checker/background.py#L69) )

but different values means an issue.

This code has run successfully:
https://github.com/searx/searx/blob/eca200404d195fcfaf29a8b30f3cc7d2c7e4c021/searx/shared/__init__.py#L15-L17

So most probably there is an issue with [searx.shared.shared_uwsgi](https://github.com/searx/searx/blob/eca200404d195fcfaf29a8b30f3cc7d2c7e4c021/searx/shared/shared_uwsgi.py) or at least with `uwsgi.ini`.

Can you have a look to your uwsgi log ?

Comment (HLFH) [reply searx/search/checker/background.py]: Ok I have finally added in`/etc/uwsgi/vassals/searx.ini`: 

```
# Cache
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
```

It works now.

Comment (HLFH) [reply searx/search/checker/background.py]: And I had this in my log.

> janv. 11 21:11:51 arch-server uwsgi[23161]: ERROR:searx.shared:uwsgi.ini configuration error, add this line to your uwsgi.ini
> janv. 11 21:11:51 arch-server uwsgi[23161]: cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1

Which is now fixed.

**Review by HLFH — CHANGES_REQUESTED** [review]
Typo ` _set_result({'status': 'unknow'})`. Ready to merge in my view afterwards.

**Review by HLFH — DISMISSED** [review]
If all the nice suggestions from the reviewers have been taken into account, this PR is good to be merged in my view.

Once merged, for the ArchLinux AUR `searx-git` package, I will:

 - add the `python-cld3-git` dependency ;
 - remove the `python-pyopenssl` dependency.

**Comment by dalf** [comment]
The last thing  I would like to check is why did you get an `error` instead of `unknown`.

Most probably this code doesn't work as expected:
https://github.com/searx/searx/blob/7f0c508598cc2197e53b877dcf4c76e25a097c4f/searx/shared/__init__.py#L19-L25

**Comment by HLFH** [comment]
> The last thing I would like to check is why did you get an `error` instead of `unknown`.
> 
> Most probably this code doesn't work as expected:
> https://github.com/searx/searx/blob/7f0c508598cc2197e53b877dcf4c76e25a097c4f/searx/shared/__init__.py#L19-L25

If I don't have this line in `/etc/uwsgi/vassals/searx.ini`, when I am quickly refreshing the `stats/checker web page`, I get:

 - sometimes error ;
 - sometimes unknown ;
 - sometimes the result itself ;
 - always (I think - or sometimes) the error log you mentioned.

So it seems it is mandatory to get this line to get things going.
If this requirement is documented properly in the searx doc, I'd say there is no problem.

**Comment by dalf** [comment]
[EDIT]

When I remove `'cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1` from uwsgi.ini, I have this log:
```
*** Operational MODE: preforking ***
added /usr/local/searx/ to pythonpath.
spawned uWSGI master process (pid: 7)
spawned uWSGI worker 1 (pid: 15, cores: 1)
spawned uWSGI worker 2 (pid: 16, cores: 1)
spawned uWSGI worker 3 (pid: 19, cores: 1)
spawned uWSGI worker 4 (pid: 24, cores: 1)
spawned 4 offload threads for uWSGI worker 1
spawned 4 offload threads for uWSGI worker 2
spawned 4 offload threads for uWSGI worker 4
spawned 4 offload threads for uWSGI worker 3
ERROR:searx.shared:uwsgi.ini configuration error, add this line to your uwsgi.ini
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
ERROR:searx.shared:uwsgi.ini configuration error, add this line to your uwsgi.ini
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
ERROR:searx.shared:uwsgi.ini configuration error, add this line to your uwsgi.ini
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
ERROR:searx.shared:uwsgi.ini configuration error, add this line to your uwsgi.ini
cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1
WSGI app 0 (mountpoint='') ready in 6 seconds on interpreter 0x55756b2761c0 pid: 15 (default app)
WSGI app 0 (mountpoint='') ready in 6 seconds on interpreter 0x55756b2761c0 pid: 16 (default app)
WSGI app 0 (mountpoint='') ready in 6 seconds on interpreter 0x55756b2761c0 pid: 24 (default app)
WSGI app 0 (mountpoint='') ready in 6 seconds on interpreter 0x55756b2761c0 pid: 19 (default app)
```

I **CAN** partially reproduce what you are describing : the checker run despite the log.

**Comment by HLFH** [comment]
@dalf Yes, the checker runs despite the log and without the recommended line, but with some various stuff shown in `stats/checker`: error/unknown/or the result itself. Shall this weird behaviour (without the line) be fixed somehow?

**Comment by dalf** [comment]
I found the issue:
https://github.com/searx/searx/blob/7f0c508598cc2197e53b877dcf4c76e25a097c4f/searx/search/checker/background.py#L83-L86

The first call to `run()` doesn't use `schedule`.

**Comment by dalf** [comment]
It should be fixed now: the checker is really disablef if uwsgi is not properly configured

This time, I think it is ready.

**Review by HLFH — APPROVED** [review]
The issue is fixed.

joe-biden-merge

**Comment by HLFH** [comment]
New `searx-git` package released on ArchLinux AUR with updated deps: https://aur.archlinux.org/packages/searx-git/.

**Comment by unixfox** [comment]
Sorry I haven't had the time to test this PR, but I just would like to know if as an instance maintainer I've to do something so that searx.space is able to use searx-checker on my instance.

**Comment by dalf** [comment]
@unixfox for now, searx-stats2 still [query the API for the searx-checker project](https://github.com/searx/searx-stats2/blob/58d0370a3304cda672fa322e6fb1e5d9a1ee8e50/searxstats/fetcher/selfreport.py#L24-L40), but with few changes it will be okay.

So as an instance maintainer, make sure to have the [checker section](https://github.com/searx/searx/blob/484dc99580a4cfdbba022209cbe896cd69b3e8b2/searx/settings.yml#L105-L131) in your settings.yml and as soon searx-stats2 is updated, searx.space will show the results for your instance.

@unixfox  Also, be to add
`cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1`
to your uwgi..ini

**Pull request merged**
